core/src/main/java/com/linecorp/armeria/client/HttpResponseDecoder.java

```diff
@@ -122,14 +123,21 @@ final boolean reserveUnfinishedResponse(int maxUnfinishedResponses) {
     }
 
     final void failUnfinishedResponses(Throwable cause) {
-        for (final Iterator<HttpResponseWrapper> iterator = responses.values().iterator();
-             iterator.hasNext();) {
-            final HttpResponseWrapper res = iterator.next();
-            // To avoid calling removeResponse by res.close(cause), remove before closing.
-            iterator.remove();
-            unfinishedResponses--;
-            res.close(cause);
+        if (wip++ > 0) {
```

Comment:
How about just using a boolean flag? such as we use `inOnNext`?

Reply:
Otherwise, how about adding `closing` flag? I think we should call failUnfinishedResponses() only once.

Reply:
That also sounds good. 👍

Reply:
By the way, the recursive remove might be called by `removeResponse()`.
Added a checking logic there too.